pkg/controller/supportbundlecollection/controller.go

```diff
@@ -667,13 +886,77 @@ func conditionExistsIgnoreLastTransitionTime(conditions []v1alpha1.SupportBundle
 	return false
 }
 
-func appendConditions(oldConditions, updatedConditions []v1alpha1.SupportBundleCollectionCondition) []v1alpha1.SupportBundleCollectionCondition {
-	newConditions := oldConditions
-	for _, c := range updatedConditions {
-		if conditionExistsIgnoreLastTransitionTime(newConditions, c) {
+func appendConditions(oldConditions, newConditions []v1alpha1.SupportBundleCollectionCondition) []v1alpha1.SupportBundleCollectionCondition {
+	finalConditions := make([]v1alpha1.SupportBundleCollectionCondition, 0)
+	newConditionMap := make(map[v1alpha1.SupportBundleCollectionConditionType]v1alpha1.SupportBundleCollectionCondition)
+	addedConditions := sets.NewString()
+	for _, condition := range newConditions {
+		newConditionMap[condition.Type] = condition
+	}
+	for _, oldCondition := range oldConditions {
+		newCondition, exists := newConditionMap[oldCondition.Type]
+		if !exists {
+			finalConditions = append(finalConditions, oldCondition)
 			continue
 		}
-		newConditions = append(newConditions, c)
+		// Use the original Condition if the only change is about lastTransition time
+		if conditionEqualsIgnoreLastTransitionTime(newCondition, oldCondition) {
+			finalConditions = append(finalConditions, oldCondition)
+		} else {
+			// Use the latest Condition.
+			finalConditions = append(finalConditions, newCondition)
+		}
+		addedConditions.Insert(string(newCondition.Type))
+	}
+	for key, newCondition := range newConditionMap {
+		if !addedConditions.Has(string(key)) {
+			finalConditions = append(finalConditions, newCondition)
+		}
+	}
+	return finalConditions
+}
+
+func getNodeKey(status *controlplane.SupportBundleCollectionNodeStatus) string {
+	// TODO: use NodeNamespace/NodeName for ExternalNode after the Namespace is added in the connection key between
+	// antrea-agent and antrea-controller.
+	return status.NodeName
+}
+
+// supportBundleCollectionStatusEqual compares two SupportBundleCollectionStatus objects. It disregards
+// the LastTransitionTime field in the status Conditions.
+func supportBundleCollectionStatusEqual(oldStatus, newStatus v1alpha1.SupportBundleCollectionStatus) bool {
+	semanticIgnoreLastTransitionTime := conversion.EqualitiesOrDie(
+		conditionSliceEqualsIgnoreLastTransitionTime,
+	)
+	return semanticIgnoreLastTransitionTime.DeepEqual(oldStatus, newStatus)
+}
+
+func conditionSliceEqualsIgnoreLastTransitionTime(as, bs []v1alpha1.SupportBundleCollectionCondition) bool {
+	sort.Slice(as, func(i, j int) bool {
```

Comment:
A way to simplify this is to always set full condition list whenever updating status, instead of only setting Completed to True but never False, and always using the same order to store status in slice.

Reply:
It is difficult to provide a full condition list any time when updating the status, for example, if the Started is false, it is meaningless to set a condition that CollectionFailure is false, and completed is false. Besides, we expect to leverage CollectionFailure=False as part of the identifier for all Nodes have uploaded the files to server. So I think the condition status is set as false when all Nodes are  succeeded to upload. But if it is set in advance ( because no failure was found at that time, but maybe happens in future), it is easier to introduce misunderstanding. 

I change the logic to sort the conditions before executing update operation to reduce the complexity, and set condition status of CollectionComplete and bundleCollected whenever updating the status, but for CollectionFailure, the status is set when a failure is reported from any Node or no failure is reported from all Nodes.

Reply:
I think the condition is there regardless of the stage of the realization, it seems you want to use "unset" to represent another state in addition to True and False. The condition value doesn't have to mean the final state, it could change based on the actual state. For example, when Started is false, it doesn't seem wrong that CollectionFailure is False and Completed is False as it just means there is no collection failure yet and the collection is not completed yet, they are obvious and redundant but not wrong and not ambiguous.

Reply:
Got it, updated.